Add different view modes for resource lists

jrauh01 opened the pull request:
This PR introduces new view modes for resource lists to enhance the user experience, following the approach used in icingadb-web. The implemantation includes three view modes:
- **minimal**: display the most essential information in one line
- **common**: like minimal, but additionally has a footer with further information
- **detailed**: like common, but the information is spreaded over three lines

If a resource does not provide sufficient information for a more detailed view, the corresponding options in the view mode switcher will not be available.

Closes #136

jrauh01 committed:
Add ViewModeSwitcher

jrauh01 committed:
Implement abstract function getIgnoredViewModes()

jrauh01 committed:
Return the matching ListItem classes

jrauh01 committed:
Implement the view modes

jrauh01 committed:
PHP style fixes

jrauh01 committed:
Icon titles for namespace list item footer

jhoxhaa reviewed (changes requested):
Could you please clarify a few things:

What are the main changes introduced in this PR?
Is there anything specific you'd like me to review or pay extra attention to?

jrauh01 commented:
I've updated the PR description to show the main changes. I think there's nothing specific that requires extra attention.

jhoxhaa commented on library/Kubernetes/Web/PodListItem.php:
```diff
@@ -35,7 +35,14 @@ class PodListItem extends BaseListItem
     protected function assembleHeader(BaseHtmlElement $header): void
```
Consider using constants for this logic, such as: const VIEW_MODE_MINIMAL = 'minimal';

jhoxhaa reviewed (changes requested):
Please use only one class instead of three (*ListItem, *ListItemMinimal, *ListItemDetailed) to handle the different levels of detail for displaying resource information.

Also use the ViewMode trait in the BaseItemList.php and BaseItemList.php instead of using it in all the other List classes.

jrauh01 committed:
Switch between view modes in one file

jrauh01 commented:
I placed the constants in ViewModeSwitcher instead, because I felt they fit better there.

jrauh01 committed:
Replace 'else if' with 'elseif'

jrauh01 committed:
Fix detail view

jrauh01 committed:
Fix referents in event detail view

jhoxhaa commented on library/Kubernetes/Web/NamespaceListItem.php:
```diff
@@ -24,15 +36,152 @@ class NamespaceListItem extends BaseListItem
+
+    protected function assembleFooter(BaseHtmlElement $footer): void
+    {
+        $resourceCount = DaemonSet::on(Database::connection())
```
Use Factory::fetchResource instead of repeating similar logic for each resource.

Also consider writing a helper method for the $resourceCount to reduce code duplication.

jrauh01 replied:
Because of the exact repeating code for each resource I would use a loop to reduce code duplication and a function to get the resources to check.

jhoxhaa commented on library/Kubernetes/Web/ViewModeSwitcher.php:
```diff
@@ -0,0 +1,238 @@
+     *
+     * @return $this
+     */
+    public function setViewMode(string $name)
```
Return type is missing.

jhoxhaa commented on library/Kubernetes/Web/ViewModeSwitcher.php:
```diff
@@ -0,0 +1,238 @@
+        return $this;
+    }
+
+    private function protectId($id)
```
Please add parameter and return type.

jrauh01 committed:
Remove redundant line

jrauh01 committed:
Add missing parameter and return type

jrauh01 committed:
Use match expressions instead of if-else

jrauh01 commented on library/Kubernetes/Web/CronJobListItem.php:
```diff
@@ -23,18 +25,41 @@ class CronJobListItem extends BaseListItem
-        );
+        $main->addHtml($this->createHeader());
+
+        if ($this->getViewMode() === ViewModeSwitcher::VIEW_MODE_DETAILED) {
```
I think using an extra variable is not necessary. Instead I can use the protected attribute $this->viewMode provided by the ViewMode trait.

jrauh01 committed:
Use ->viewMode instead of ->getViewMode()

jrauh01 committed:
Remove superfluous if statement

jrauh01 committed:
Simplify namespace's caption and footer

jrauh01 committed:
Hide view mode switcher

Hide view mode switcher in the Web UI if all view modes are ignored or
only one view mode is not ignored.

jrauh01 committed:
Format docs

jhoxhaa commented on library/Kubernetes/Common/ViewMode.php:
```diff
@@ -0,0 +1,35 @@
+<?php
+
+/* Icinga for Kubernetes Web | (c) 2025 Icinga GmbH | GPLv2 */
```
We are now using the AGPLv3 license. Please ensure this change is applied across all classes.

jrauh01 committed:
Fix license headers

jrauh01 committed:
Pass db via argument

Pass database via argument to fetchResource() instead of creating a new
connection for each loop pass what can lead to an 'Too many
connections' error.

jrauh01 committed:
Make getIgnoredViewModes() not abstract anymore

jrauh01 committed:
Add function to canonicalize resource kinds

jrauh01 committed:
Add default traits for list item parts

jrauh01 committed:
Adjust function names to canonicalized kind names

jrauh01 committed:
Use default traits for item list parts

jrauh01 committed:
Add styles for list items

jrauh01 committed:
Extend `ItemList` to display resources

jrauh01 committed:
Use new list implementation for conditions

jrauh01 committed:
Implement `ItemRenderer` ...

... to provide a base to render resources.

jrauh01 committed:
Extend `BaseResourceRenderer` to render resources

jrauh01 committed:
Use `ResourceList`

jrauh01 committed:
Full width for list items in detail view

jrauh01 committed:
Add default item layout for resources

Show footer instead of caption for `Secrets`.

jrauh01 committed:
Adjust styles for new list implementation

jrauh01 committed:
Remove old list implementation

jrauh01 committed:
Add list items as detail view header

jrauh01 committed:
Adjust size for pv and pvc state icons

jrauh01 committed:
Standardize empty state messages in detail view

jrauh01 committed:
Adjust `createDetailUrl()` factory function

Add SidecarContainers and InitContainers and return pvc for
PersistentVolumeClaims

lippserd commented on library/Kubernetes/View/CronJobRenderer.php:
```diff
@@ -1,82 +1,73 @@
 
-        if (isset($this->item->last_successful_time)) {
-            $lastSuccessfulTime = $this->item->last_successful_time->format('Y-m-d H:i:s');
+        if (isset($item->last_successful_time)) {
+            $lastSuccessfulTime = $item->last_successful_time->format('Y-m-d H:i:s');
         } else {
             $lastSuccessfulTime = $this->translate('None');
         }
```
```suggestion
        $lastSuccessfulTime = $item-> last_successful_time?->format('Y-m-d H:i:s') ?? $this->translate('None');
```

Newline remaining from the two if blocks can be removed then.

lippserd commented on library/Kubernetes/Web/Conditions.php:
```diff
@@ -31,53 +28,10 @@ protected function assemble(): void
 
-        $listItems = [];
-
         $conditions = yield_iterable($this->getConditions());
```
`$this->getConditions())` can be directly used now in the `new ConditionList` call.

lippserd commented on library/Kubernetes/Web/Factory.php:
```diff
@@ -27,16 +27,23 @@
 {
+    public static function canonicalizeKind(string $kind): string
+    {
+        if ($kind === 'pvc') {
+            return 'persistentvolumeclaim';
+        }
+        return strtolower(str_replace(['_', '-'], '', $kind));
```
```suggestion
        $kind = strtolower($kind);

        return match ($kind) {
            'pvc'   => 'persistentvolumeclaim',
            default => str_replace(['_', '-'], '', $kind),
        };
```

lippserd commented on library/Kubernetes/Web/Factory.php:
```diff
@@ -27,16 +27,23 @@
     public static function createIcon(string $kind): ?ValidHtml
     {
-        $kind = strtolower(str_replace(['_', '-'], '', $kind));
+        $kind = self::canonicalizeKind($kind);
```
```suggestion
        $kind = static::canonicalizeKind($kind);
```

lippserd commented on library/Kubernetes/Web/Factory.php:
```diff
@@ -60,86 +67,6 @@ public static function createIcon(string $kind): ?ValidHtml
-
     public static function createModel(string $kind): ?Model
     {
         $kind = strtolower(str_replace(['_', '-'], '', $kind));
```
```suggestion
        $kind = static::canonicalizeKind($kind);
```

lippserd commented on library/Kubernetes/Web/Factory.php:
```diff
@@ -27,16 +27,23 @@
-use ipl\Web\Widget\EmptyState;
 use ipl\Web\Widget\Icon;
 
 abstract class Factory
```
In addition to the requested changes this class, please use `self::canonicalizeKind($kind)` everywhere, omit comma in the last match case and use direct return with match where appropriate.

lippserd commented on library/Kubernetes/Web/ItemList/ResourceList.php:
```diff
@@ -0,0 +1,142 @@
+     */
+    public function setViewMode(string $mode): self
+    {
+        switch ($mode) {
+            case 'minimal':
+                $this->setItemLayoutClass(MinimalItemLayout::class);
+
+                break;
+            case 'detailed':
+                $this->setItemLayoutClass(DetailedItemLayout::class);
+
+                break;
+            case 'common':
+                $this->setItemLayoutClass(ResourceDefaultItemLayout::class);
+
+                break;
+            default:
+                throw new InvalidArgumentException('Invalid view mode');
+        }
```
```suggestion
        $class = match ($mode) {
            'minimal'  => MinimalItemLayout::class,
            'detailed' => DetailedItemLayout::class,
            'common'   => ResourceDefaultItemLayout::class,
            default    => throw new InvalidArgumentException('Invalid view mode')
        };

        $this->setItemLayoutClass($class);
```

lippserd replied:
Has to be adjusted again once $mode is an enum.

jrauh01 replied:
Do I have to store the class in an extra variable or can I just do the following:
```php
$this->setItemLayoutClass(match ($mode) {
    ViewModeSwitcher::VIEW_MODE_MINIMAL  => MinimalItemLayout::class       
    ViewModeSwitcher::VIEW_MODE_COMMON   => DetailedItemLayout::class,
    ViewModeSwitcher::VIEW_MODE_DETAILED => ResourceDefaultItemLayout::class,
    default                              => throw new InvalidArgumentException('Invalid view mode')
});
```

lippserd commented on library/Kubernetes/Web/ViewModeSwitcher.php:
```diff
@@ -0,0 +1,245 @@
+use ipl\Web\Common\FormUid;
+use ipl\Web\Widget\IcingaIcon;
+
+class ViewModeSwitcher extends Form
```
* `use Translation`
* Use features up to PHP 8.2., e.g., property types, view mode enum, static return type.

jrauh01 commented on library/Kubernetes/View/ConditionRenderer.php:
```diff
@@ -0,0 +1,60 @@
+class ConditionRenderer implements ItemRenderer
+{
+    /** @var callable */
+    protected $getVisualFunc;
```
Property cannot have type 'callable'

lippserd replied:
True, make it `Closure` then ;).

jrauh01 committed:
Lowercase container links functions

jrauh01 committed:
Use new container icon
